### PR TITLE
Lab 5

### LINKED ISSUE #299: Lab 5
- Implement stringer interface in type IPAddr
- Add Unit Tests for 100% test coverage

### PR DESCRIPTION
**Pull request by joshcarp**
Fixes #299

Review of colleague's PR #

#### Changes proposed in this PR:

- Implement stringer interface in type IPAddr
- Add Unit Tests for 100% test coverage

### CODE REVIEW

File: 05_stringer/joshcarp/IPAddr.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package main
+
+import "fmt"
+
+type IPAddr struct {
```
Comment (marksomething) [inline 05_stringer/joshcarp/IPAddr.go:5]: Probably trying to get too clever here, [CIDR](https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing) means that there isn't really fixed purpose for the octets.

File: 05_stringer/joshcarp/IPAddr_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	os.Stdout = w
+	return r, w
+}
+
+// captureStop copies a file buffer and returns a string of the file
+func captureStop(r io.Reader, w io.Closer) string {
+	var buf bytes.Buffer
+	w.Close()
+	os.Stdout = stdout
+	_, err := io.Copy(&buf, r)
```
Comment (marksomething) [inline 05_stringer/joshcarp/IPAddr_test.go:49]: combining the execution of a function and it's test for nil error seems to be idiomatic in go

File: 05_stringer/joshcarp/IPAddr_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+func TestStringer(t *testing.T) {
+	for _, test := range tests {
+		require.Equal(t, test.expected, test.input.String())
+	}
+}
+
+// captureStart diverts stdio to another file object
+func captureStart() (io.Reader, io.Closer) {
```
Comment (marksomething) [inline 05_stringer/joshcarp/IPAddr_test.go:38]: A lot of people are using `fmt.Fprintln()` in their main function to make this process much simpler. This is cool though.

File: 05_stringer/joshcarp/IPAddr_test.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+	"github.com/stretchr/testify/require"
+)
+
+var stdout = os.Stdout
+
+var tests = []struct {
+	input    IPAddr
+	expected string
+}{
```
Comment (marksomething) [inline 05_stringer/joshcarp/IPAddr_test.go:17]: Most people are testing (and handling) partial/empty IPAddr ( eg. IPAddr{127}, IPAddr{} )

File: 05_stringer/joshcarp/IPAddr.go (hunk 1 above)
Comment (marksomething) [reply 05_stringer/joshcarp/IPAddr.go:5]: `[4]byte` rather than a struct is appropriate

**Review by marksomething — CHANGES_REQUESTED** [review]

File: 05_stringer/joshcarp/IPAddr_test.go (hunk 2 above)
Comment (joshcarp) [reply 05_stringer/joshcarp/IPAddr_test.go:49]: i dont know what this means

Comment (joshcarp) [reply 05_stringer/joshcarp/IPAddr_test.go:49]: you mean 
``` if _, err := io.Copy(&buf, r); err != nil{```
?

File: 05_stringer/joshcarp/IPAddr.go (hunk 1 above)
Comment (joshcarp) [reply 05_stringer/joshcarp/IPAddr.go:5]: Yeah well my point was this was my approach before I knew that ```[4]byte``` could be used, and like in the spirit of not copying peoples code and all

File: 05_stringer/joshcarp/IPAddr_test.go (hunk 2 above)
Comment (marksomething) [reply 05_stringer/joshcarp/IPAddr_test.go:49]: yes

File: 05_stringer/joshcarp/IPAddr.go (hunk 1 above)
Comment (marksomething) [reply 05_stringer/joshcarp/IPAddr.go:5]: I'm not quite sure how to respond to this, i agree it's a good approach to develop your code before looking at others (i'm doing that too). It's also a normal code review process to receive feedback and improve code in response to feedback (or argue the case why you think your approach is better :) ).

File: 05_stringer/joshcarp/IPAddr_test.go (hunk 2 above)
Comment (juliaogris) [reply 05_stringer/joshcarp/IPAddr_test.go:49]: yup agree with @marksomething - please fix.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Please add link to your review of colleague's PR.

**Comment by anzdaddy** [comment]
Please close merge or close.

**Pull request closed without merging**